### ISSUE TITLE
Self Defined Logloss does not match with built-in loss function

### ISSUE THREAD

**Issue by celticsalan** [catboost/catboost#1847]
Problem: Self Defined Logloss does not match with built in loss function
catboost version: 0.26.1
Operating System: Linux
CPU: 3

Hi,

I have used your user-defined log-loss objective (defined here "https://catboost.ai/docs/concepts/python-usages-examples.html") and applied it on a simple dataset. But I can't reproduce the same results that the built-in logloss produced. To be more specific, I can generate same results when number of iterations is small, but when I increase the number of iterations, the results gets far off. I was wondering if this is because of the built-n logloss function is saved in an _catboost.so file, so that I can't create same results using self-written function in python.

Here is a minimum working example.

#from catboost import CatBoostClassifier
# Initialize data
cat_features = [0, 1]
train_data = [["a", "b", 1, 4, 5, 6],
              ["a", "b", 4, 5, 6, 7],
              ["c", "d", 30, 40, 50, 60]]
train_labels = [1, 1, -1]
eval_data = [["a", "b", 2, 4, 6, 8],
             ["a", "d", 1, 4, 50, 60]]

class LoglossObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)
        
        result = []
        for index in range(len(targets)):
            e = np.exp(approxes[index])
            p = e / (1 + e)
            der1 = targets[index] - p
            der2 = -p * (1 - p)

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index]

            result.append((der1, der2))
        return result

new_cat_params ={'iterations': 1000,
                 'learning_rate': 0.01,
                 'depth': 10,
                 'border_count': 20,
                 'thread_count': 3,
                 'random_seed': 1371,
                 'verbose': False, 
                 'cat_features': cat_features
                }

# Model with Built-in log-loss
model = CatBoostClassifier(**new_cat_params)
aaa = model.fit(train_data, train_labels)
print("Built-in Logloss: ", aaa.predict_proba(eval_data)[:,1])

# Model with Customized Loss function
model = CatBoostClassifier(**new_cat_params,
                              loss_function=LoglossObjective(),
                              eval_metric = "Logloss")
aaa = model.fit(train_data, train_labels)
print("Self-Written Logloss: ", aaa.predict_proba(eval_data)[:,1])
